src/LinuxTracing/PerfEventOpen.cpp

```diff
@@ -193,4 +193,45 @@ int tracepoint_event_open(const char* tracepoint_category, const char* tracepoin
   return generic_event_open(&pe, pid, cpu);
 }
 
+int tracepoint_with_callchain_event_open(const char* tracepoint_category,
+                                         const char* tracepoint_name, pid_t pid, int32_t cpu,
+                                         uint16_t stack_dump_size) {
+  int tp_id = GetTracepointId(tracepoint_category, tracepoint_name);
+  if (tp_id == -1) {
+    return -1;
+  }
+  perf_event_attr pe = generic_event_attr();
+  pe.type = PERF_TYPE_TRACEPOINT;
+  pe.config = tp_id;
+  pe.sample_type |= PERF_SAMPLE_CALLCHAIN | PERF_SAMPLE_RAW;
+  // TODO(b/239003729): Read this from /proc/sys/kernel/perf_event_max_stack
+  pe.sample_max_stack = 127;
+  pe.exclude_callchain_kernel = true;
+
+  // Also capture a small part of the stack and the registers to allow patching the callers of
+  // leaf functions. This is done by unwinding the first two frame using DWARF.
+  pe.sample_type |= PERF_SAMPLE_REGS_USER | PERF_SAMPLE_STACK_USER;
+  pe.sample_regs_user = SAMPLE_REGS_USER_ALL;
+  pe.sample_stack_user = stack_dump_size;
+
+  return generic_event_open(&pe, pid, cpu);
+}
+
+int tracepoint_with_stack_event_open(const char* tracepoint_category, const char* tracepoint_name,
+                                     pid_t pid, int32_t cpu, uint16_t stack_dump_size) {
+  int tp_id = GetTracepointId(tracepoint_category, tracepoint_name);
+  if (tp_id == -1) {
+    return -1;
+  }
+  perf_event_attr pe = generic_event_attr();
+  pe.type = PERF_TYPE_TRACEPOINT;
+  pe.config = tp_id;
+  pe.sample_type |= PERF_SAMPLE_REGS_USER | PERF_SAMPLE_STACK_USER | PERF_SAMPLE_RAW;
+  pe.sample_regs_user = SAMPLE_REGS_USER_ALL;
+
```

Comment:
NIt: omit the empty line. The stack size corresponds to the "PERF_SAMPLE_STACK_USER".